Converted TimerService to C++

mandrek44 opened the pull request:
I converted TimerService to C++ implementation.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pw-sat2/pwsat2obc/53)
<!-- Reviewable:end -->

decu69 commented:
Reviewed 12 of 12 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53)*
<!-- Sent from Reviewable.io -->

Novakov commented:
Reviewed 10 of 12 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[libs/time/include/time/timer.h, line 21 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXNMTtmQr9MDQ04q80R:-KXNMTtmQr9MDQ04q80S:bo2lr6a) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/b1b318be05aed34206cf4b68132466c589b0fd78/libs/time/include/time/timer.h#L21)):*
> ```objc
> {
>     namespace time
>     {
> ```

We need to think about namespaces. It's easy to get too deep structure

---

*[libs/time/include/time/timer.h, line 89 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXNNV9ZsQHZ0C8VXDgv:-KXNNV9ZsQHZ0C8VXDgw:bnx9p6k) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/b1b318be05aed34206cf4b68132466c589b0fd78/libs/time/include/time/timer.h#L89)):*
> ```objc
>              * @return True if this time snapshot is considered smaller than the other one, false otherwise.
>              */
>             inline bool LessThan(struct TimeSnapshot other)
> ```

@zaq32 Any objectections on overloading comparision operators for time-related types?

---

*[libs/time/include/time/timer.h, line 165 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXNMxhBywfZMAa3NjoR:-KXNMxhCD93ACQkdqy8J:b-q8iufy) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/b1b318be05aed34206cf4b68132466c589b0fd78/libs/time/include/time/timer.h#L165)):*
> ```objc
>                  * @brief Pointer to file system object that is used to save/restore timer state.
>                  */
>                 FileSystem* FileSystemObject;
> ```

Fields should go after methods, as they are more important

---

*[libs/time/include/time/timer.h, line 183 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXNOXkCB5_1MnPW8qjK:-KXNOXkCB5_1MnPW8qjL:b96uefk) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/b1b318be05aed34206cf4b68132466c589b0fd78/libs/time/include/time/timer.h#L183)):*
> ```objc
>                     TimePassedCallbackType timePassedCallback, //
>                     void* timePassedCallbackContext,           //
>                     FileSystem* fileSystem                     //
> ```

Dependencies like `FileSystem` should be passed to constructur and by reference

---

*[libs/time/include/time/timer.h, line 193 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXNOsHXjuibmrw8N-EZ:-KXNOsHXjuibmrw8N-E_:bcv65wi) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/b1b318be05aed34206cf4b68132466c589b0fd78/libs/time/include/time/timer.h#L193)):*
> ```objc
>                  * @return True on success, false otherwise.
>                  */
>                 bool GetCurrentTime(TimeSpan* currentTime);
> ```

This method effectively returns two values: success flag and time. It will be wise to pack it together in some sort of structure (maybe some template to use is other methods?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53)*
<!-- Sent from Reviewable.io -->

mandrek44 commented:
Review status: 8 of 12 files reviewed at latest revision, 5 unresolved discussions.

---

*[libs/time/include/time/timer.h, line 21 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXNMTtmQr9MDQ04q80R:-KXbBbQ1NnCQmC_3N5wL:b5rd9jv) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/b1b318be05aed34206cf4b68132466c589b0fd78/libs/time/include/time/timer.h#L21)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary>
> We need to think about namespaces. It's easy to get too deep structure
</details>
I think that all obc related classes should be in `obc` namespace, so we don't collide with anything else (like `time`).

---

*[libs/time/include/time/timer.h, line 183 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXNOXkCB5_1MnPW8qjK:-KXbBLu9tbHn3QLqOwwe:bqe7gy1) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/b1b318be05aed34206cf4b68132466c589b0fd78/libs/time/include/time/timer.h#L183)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary>
> Dependencies like `FileSystem` should be passed to constructur and by reference
</details>
Should it be const reference? (like `const FileSystem& fileSystem`)?

---

*[libs/time/include/time/timer.h, line 193 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXNOsHXjuibmrw8N-EZ:-KXbEchQnpaU_4gSntzF:b-sgxm65) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/b1b318be05aed34206cf4b68132466c589b0fd78/libs/time/include/time/timer.h#L193)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary>
> This method effectively returns two values: success flag and time. It will be wise to pack it together in some sort of structure (maybe some template to use is other methods?)
</details>
This seems like common pattern - do we have an existing implementation?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53)*
<!-- Sent from Reviewable.io -->

zaq32 commented:
Now that we are using c++ it is worth considering using std::chrono::milliseconds type instead of TimeSpan.

Also if the std::chrono would not be used I'd recommend defining basic operators arithmetic & logic operators for TimeSpan & TimePoint types

---

Reviewed 6 of 7 files at r2.
Review status: 11 of 12 files reviewed at latest revision, 10 unresolved discussions.

---

*[libs/time/timer.cpp, line 23 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXgOD6QJVcdB5TzVAMS:-KXgOD6RYpmqqlC2ytLq:bzcsb9e) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/0ffa8b20990830c3836f1845389b833a5d8ae63e/libs/time/timer.cpp#L23)):*
> ```cpp
>     FileSystemObject = &fileSystem;
>     NotificationTime = TimeSpanFromMilliseconds(0ull);
>     PersistanceTime = TimeSpanFromMilliseconds(0ull);
> ```

Use class initializer list instead of  direct assignments

---

*[libs/time/include/time/timer.h, line 89 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXNNV9ZsQHZ0C8VXDgv:-KXgPsNxvA4PgJyBNCnv:b-7hzvg) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/b1b318be05aed34206cf4b68132466c589b0fd78/libs/time/include/time/timer.h#L89)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary><blockquote>

@zaq32 Any objectections on overloading comparision operators for time-related types?
</blockquote></details>

No. In fact now that we do have use c++ it would be preferable to define at least <, ==, != operators for this type and for TimeSpan & TimePoint

---

*[libs/time/include/time/timer.h, line 183 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXNOXkCB5_1MnPW8qjK:-KXgQTOrhBvXeT81V3vn:bdcsnyu) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/b1b318be05aed34206cf4b68132466c589b0fd78/libs/time/include/time/timer.h#L183)):*
<details><summary><i>Previously, mandrek44 (Mandro) wrote…</i></summary><blockquote>

Should it be const reference? (like `const FileSystem& fileSystem`)?
</blockquote></details>

reference to const FileSystem may be temping but it would be difficult to get right the FileSystem interface itself

---

*[libs/time/include/time/timer.h, line 193 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXNOsHXjuibmrw8N-EZ:-KXgR9UImOb7QRmlvH1B:bpz6glc) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/b1b318be05aed34206cf4b68132466c589b0fd78/libs/time/include/time/timer.h#L193)):*
<details><summary><i>Previously, mandrek44 (Mandro) wrote…</i></summary><blockquote>

This seems like common pattern - do we have an existing implementation?
</blockquote></details>

No yet.

---

*[libs/time/include/time/timer.h, line 278 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXgTYgBcd2fGKROBpeD:-KXgTYgBcd2fGKROBpeE:b3aqg0y) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/0ffa8b20990830c3836f1845389b833a5d8ae63e/libs/time/include/time/timer.h#L278)):*
> ```objc
>                  * @brief Time period between the subsequent mission time notifications.
>                  */
>                 const TimeSpan NotificationPeriod = {TIMER_NOTIFICATION_PERIOD};
> ```

since this value is hard coded I'd mark it as static constexpr instead of making it const member.

---

*[libs/time/include/time/timer.h, line 283 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXgTig6TYWEzqrr7-Nx:-KXgTig6TYWEzqrr7-Ny:b3aqg0y) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/0ffa8b20990830c3836f1845389b833a5d8ae63e/libs/time/include/time/timer.h#L283)):*
> ```objc
>                  * @brief Time period between subsequent timer state saves.
>                  */
>                 const TimeSpan SavePeriod = {TIMER_SAVE_PERIOD};
> ```

since this value is hard coded I'd mark it as static constexpr instead of making it const member.

---

*[libs/time/include/time/timer.h, line 287 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXgRKcWw-_RqDq6Qv_c:-KXgRKcXZ-Tsr_X8JOd2:b52gqw0) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/0ffa8b20990830c3836f1845389b833a5d8ae63e/libs/time/include/time/timer.h#L287)):*
> ```objc
>                 const char* const File0 = "/TimeState.0";
>                 const char* const File1 = "/TimeState.1";
>                 const char* const File2 = "/TimeState.2";
> ```

these three fields are not supposed to be changed or instance specific therefore mark them as static. 

---

*[src/main.cpp, line 176 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXgRdIg2TMSfy7ISdCZ:-KXgRdIg2TMSfy7ISdC_:bwo3qm3) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/0ffa8b20990830c3836f1845389b833a5d8ae63e/src/main.cpp#L176)):*
> ```cpp
>     ClearState(obc);
> 
>     if (!obc->timeProvider.Initialize(NULL, NULL))
> ```

use nullptr instead of NULL

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-:-KXgSq2buuc3n8Cya2B9:b-76qr5e)*
<!-- Sent from Reviewable.io -->

zaq32 commented:
Reviewed 1 of 7 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53)*
<!-- Sent from Reviewable.io -->

mandrek44 commented:
Agree but I'd like to do this as seperate work / pull request

---

Review status: 2 of 13 files reviewed at latest revision, 10 unresolved discussions.

---

*[libs/time/include/time/timer.h, line 21 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXNMTtmQr9MDQ04q80R:-KY0KK8lasO8lqtIdEst:bvakhyi) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/b1b318be05aed34206cf4b68132466c589b0fd78/libs/time/include/time/timer.h#L21)):*
<details><summary><i>Previously, mandrek44 (Mandro) wrote…</i></summary><blockquote>

I think that all obc related classes should be in `obc` namespace, so we don't collide with anything else (like `time`).
</blockquote></details>

I changed it to services per our discussions

---

*[libs/time/include/time/timer.h, line 89 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXNNV9ZsQHZ0C8VXDgv:-KY03E32WRvVB7ZUIhik:b-9rt1oi) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/b1b318be05aed34206cf4b68132466c589b0fd78/libs/time/include/time/timer.h#L89)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

No. In fact now that we do have use c++ it would be preferable to define at least <, ==, != operators for this type and for TimeSpan & TimePoint
</blockquote></details>

I haven't ported the `TimePoint` and `TimeSpan` yet - I'd first finish this bit before starting new ;-)
It seems to me that if I overload the operators the `LessThan` and `Eqaul` methods can be removed, right?

---

*[libs/time/include/time/timer.h, line 193 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXNOsHXjuibmrw8N-EZ:-KY0KQrSKpQg9Fu4MQql:brszt72) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/b1b318be05aed34206cf4b68132466c589b0fd78/libs/time/include/time/timer.h#L193)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

No yet.
</blockquote></details>

I added the `Option` class

---

*[libs/time/include/time/timer.h, line 278 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXgTYgBcd2fGKROBpeD:-KY0KTxyMIYt-fzWlKSx:b2zbgq4) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/0ffa8b20990830c3836f1845389b833a5d8ae63e/libs/time/include/time/timer.h#L278)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

since this value is hard coded I'd mark it as static constexpr instead of making it const member.
</blockquote></details>

Compiler complained that it should be constexpr, but then it complained that it's undefined reference. I admin I got lost there ;-)

---

*[src/main.cpp, line 176 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXgRdIg2TMSfy7ISdCZ:-KXr10ymmU9AvcwN1gJo:b-2h35xj) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/0ffa8b20990830c3836f1845389b833a5d8ae63e/src/main.cpp#L176)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

use nullptr instead of NULL
</blockquote></details>

NULLs are used all over the timer.cpp - should I also changed all of them?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-:-KY0K6YOSPHlwQvC9VeZ:b-a6yzqq)*
<!-- Sent from Reviewable.io -->

mandrek44 commented:
Review status: 2 of 13 files reviewed at latest revision, 10 unresolved discussions.

---

*[libs/time/include/time/timer.h, line 283 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXgTig6TYWEzqrr7-Nx:-KY0L-1aH1ZYIxuRWRjP:b2zbgq4) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/0ffa8b20990830c3836f1845389b833a5d8ae63e/libs/time/include/time/timer.h#L283)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

since this value is hard coded I'd mark it as static constexpr instead of making it const member.
</blockquote></details>

Compiler complained that it should be constexpr, but then it complained that it's undefined reference. I admin I got lost there ;-)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53)*
<!-- Sent from Reviewable.io -->

mandrek44 commented:
Review status: 2 of 13 files reviewed at latest revision, 10 unresolved discussions.

---

*[libs/time/timer.cpp, line 23 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXgOD6QJVcdB5TzVAMS:-KY0L8dZBuU1bVOctWpq:ba7hmqd) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/0ffa8b20990830c3836f1845389b833a5d8ae63e/libs/time/timer.cpp#L23)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

Use class initializer list instead of  direct assignments
</blockquote></details>

Done, but clang-format has reduced the lines

---

*[libs/time/include/time/timer.h, line 165 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXNMxhBywfZMAa3NjoR:-KY0LLL4M40E8rGlwwj9:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/b1b318be05aed34206cf4b68132466c589b0fd78/libs/time/include/time/timer.h#L165)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary><blockquote>

Fields should go after methods, as they are more important
</blockquote></details>

Done.

---

*[libs/time/include/time/timer.h, line 183 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXNOXkCB5_1MnPW8qjK:-KY0LMVRwHtWW06ofBPW:b32w1t) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/b1b318be05aed34206cf4b68132466c589b0fd78/libs/time/include/time/timer.h#L183)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

reference to const FileSystem may be temping but it would be difficult to get right the FileSystem interface itself
</blockquote></details>

Ok

---

*[libs/time/include/time/timer.h, line 287 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXgRKcWw-_RqDq6Qv_c:-KY0L434nGxeHeNlhnGG:b-mvn7bx) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/0ffa8b20990830c3836f1845389b833a5d8ae63e/libs/time/include/time/timer.h#L287)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

these three fields are not supposed to be changed or instance specific therefore mark them as static. 
</blockquote></details>

A made them static constexpr

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53)*
<!-- Sent from Reviewable.io -->

Novakov commented:
Reviewed 2 of 7 files at r2, 11 of 11 files at r3.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[libs/base/Include/utils.h, line 58 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KY0NFBL_ZLfRnpOhVeS:-KY0NFBL_ZLfRnpOhVeT:b-htk9vs) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/c5633e33364379d731545fdcef9a65b69fe2c0cf/libs/base/Include/utils.h#L58)):*
> ```objc
> 
> template <class T>
> class Option
> ```

Undocumented

---

*[libs/base/Include/utils.h, line 68 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KY0NiqwPtaYuVLC0cnN:-KY0NiqwPtaYuVLC0cnO:b59eryw) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/c5633e33364379d731545fdcef9a65b69fe2c0cf/libs/base/Include/utils.h#L68)):*
> ```objc
>     static Option<T> Some(T &result)
>     {
>         return Option<T>(true, result);
> ```

+@zaq32 Would it be better to pass result by value and use std::move?

---

*[libs/time/timer.cpp, line 23 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXgOD6QJVcdB5TzVAMS:-KY0O5bLcpCl4zCc8yji:bo688za) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/0ffa8b20990830c3836f1845389b833a5d8ae63e/libs/time/timer.cpp#L23)):*
<details><summary><i>Previously, mandrek44 (Mandro) wrote…</i></summary><blockquote>

Done, but clang-format has reduced the lines
</blockquote></details>

You can use `//` at the end of line to force specific formatting. We use that trick in few places

---

*[libs/time/include/time/timer.h, line 89 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXNNV9ZsQHZ0C8VXDgv:-KY0ORVCmz_dRGkNlsuU:bexhfqg) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/b1b318be05aed34206cf4b68132466c589b0fd78/libs/time/include/time/timer.h#L89)):*
<details><summary><i>Previously, mandrek44 (Mandro) wrote…</i></summary><blockquote>

I haven't ported the `TimePoint` and `TimeSpan` yet - I'd first finish this bit before starting new ;-)
It seems to me that if I overload the operators the `LessThan` and `Eqaul` methods can be removed, right?
</blockquote></details>

Right, remove them and define operators.

---

*[libs/time/include/time/timer.h, line 193 at r1](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXNOsHXjuibmrw8N-EZ:-KY0PcZEqz3aNb1SNp1J:bfkdjt6) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/b1b318be05aed34206cf4b68132466c589b0fd78/libs/time/include/time/timer.h#L193)):*
<details><summary><i>Previously, mandrek44 (Mandro) wrote…</i></summary><blockquote>

I added the `Option` class
</blockquote></details>

Update documenation to match new signature

---

*[libs/time/include/time/timer.h, line 189 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KY0Plog6ZgD3I6v8TRQ:-KY0Plog6ZgD3I6v8TRR:b-bkd134) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/c5633e33364379d731545fdcef9a65b69fe2c0cf/libs/time/include/time/timer.h#L189)):*
> ```objc
>              * This value is protected by the timerLock semaphore.
>              */
>             TimeSpan CurrentTime;
> ```

I think that these fields could be private

---

*[src/obc.cpp, line 6 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KY0P-kDFAUpcGj8biQk:-KY0P-kDFAUpcGj8biQl:b-bldz3f) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/c5633e33364379d731545fdcef9a65b69fe2c0cf/src/obc.cpp#L6)):*
> ```cpp
> OBC::OBC()
>     : timeProvider(fs), I2C(&I2CBuses[I2C_SYSTEM_BUS].ErrorHandling.Base, &I2CBuses[I2C_PAYLOAD_BUS].ErrorHandling.Base),
>       Communication(*I2C.Bus), terminal(this->IO)
> ```

Comments at the end of lines forced more readable/mergeable format - bring them back :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53)*
<!-- Sent from Reviewable.io -->

zaq32 commented:
Review status: 12 of 13 files reviewed at latest revision, 12 unresolved discussions.

---

*[libs/base/Include/utils.h, line 68 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KY0NiqwPtaYuVLC0cnN:-KY0QU_W2KSCZzP475da:b-rzk7kj) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/c5633e33364379d731545fdcef9a65b69fe2c0cf/libs/base/Include/utils.h#L68)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary><blockquote>

+@zaq32 Would it be better to pass result by value and use std::move?
</blockquote></details>

I would rather use universal reference + std::move 

---

*[libs/time/timer.cpp, line 38 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KY0QeUgtX9dXaiDGXBo:-KY0QeUgtX9dXaiDGXBp:bepkzch) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/c5633e33364379d731545fdcef9a65b69fe2c0cf/libs/time/timer.cpp#L38)):*
> ```cpp
> }
> 
> inline bool operator>=(const TimeSnapshot& left, const TimeSnapshot& right)
> ```

As those operators are declared inline and they are defined inside the source file (as opposed to header file) they will be unavailable to anyone outside this translation unit

---

*[libs/time/include/time/timer.h, line 278 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXgTYgBcd2fGKROBpeD:-KY0SwsD93bMJkFe3jsn:bplejco) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/0ffa8b20990830c3836f1845389b833a5d8ae63e/libs/time/include/time/timer.h#L278)):*
<details><summary><i>Previously, mandrek44 (Mandro) wrote…</i></summary><blockquote>

Compiler complained that it should be constexpr, but then it complained that it's undefined reference. I admin I got lost there ;-)
</blockquote></details>

did you make mark it static as well? 

---

*[libs/time/include/time/timer.h, line 283 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXgTig6TYWEzqrr7-Nx:-KY0SrZsrELGyUa_sGYp:bplejco) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/0ffa8b20990830c3836f1845389b833a5d8ae63e/libs/time/include/time/timer.h#L283)):*
<details><summary><i>Previously, mandrek44 (Mandro) wrote…</i></summary><blockquote>

Compiler complained that it should be constexpr, but then it complained that it's undefined reference. I admin I got lost there ;-)
</blockquote></details>

did you make mark it static as well? 

---

*[src/main.cpp, line 176 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXgRdIg2TMSfy7ISdCZ:-KY0Sz1VyFkYJUym5VxS:b-1ih49x) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/0ffa8b20990830c3836f1845389b833a5d8ae63e/src/main.cpp#L176)):*
<details><summary><i>Previously, mandrek44 (Mandro) wrote…</i></summary><blockquote>

NULLs are used all over the timer.cpp - should I also changed all of them?
</blockquote></details>

NULL comes from C and has not type which makes it easy to misuse due to the hack in the C & C++ standards. nullptr on the other hand has its own type and cannot be misused therefore it preferable. Both are still fine. In new code prefer using nullptr in already existing one I leave it up to you

---

*[src/mission.cpp, line 44 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KY0TeQxWJBePh94PRSJ:-KY0TeQxWJBePh94PRSK:b-po62a) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/c5633e33364379d731545fdcef9a65b69fe2c0cf/src/mission.cpp#L44)):*
> ```cpp
>     UNREFERENCED_PARAMETER(param);
> 
>     state->Time = Main.timeProvider.GetCurrentMissionTime().Value;
> ```

Update the time only if it actually has been successfully obtained.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53)*
<!-- Sent from Reviewable.io -->

zaq32 commented:
Review status: 12 of 13 files reviewed at latest revision, 11 unresolved discussions.

---

*[libs/base/Include/utils.h, line 68 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KY0NiqwPtaYuVLC0cnN:-KY0Vi_L32iJ-yoAcBQk:b-d9xnss) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/c5633e33364379d731545fdcef9a65b69fe2c0cf/libs/base/Include/utils.h#L68)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

I would rather use universal reference + std::move 
</blockquote></details>

* std::forward not std::move

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53)*
<!-- Sent from Reviewable.io -->

zaq32 commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53)*
<!-- Sent from Reviewable.io -->

mandrek44 commented:
Review status: 2 of 18 files reviewed at latest revision, 12 unresolved discussions.

---

*[libs/base/Include/utils.h, line 58 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KY0NFBL_ZLfRnpOhVeS:-KYVSzKRMbsklHJ96oEP:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/c5633e33364379d731545fdcef9a65b69fe2c0cf/libs/base/Include/utils.h#L58)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary><blockquote>

Undocumented
</blockquote></details>

Done.

---

*[libs/base/Include/utils.h, line 68 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KY0NiqwPtaYuVLC0cnN:-KY_54sXcDpIlJJMrTBW:b-cvnv3a) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/c5633e33364379d731545fdcef9a65b69fe2c0cf/libs/base/Include/utils.h#L68)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

* std::forward not std::move
</blockquote></details>

I used `std::move` so that `Option<T>` can be constructed from rvalue references - which makes sense as long as it's used for returning value.  I had problem adopting `std::forward`, because for lvalue references passed to factory method (the one in global scope) the type argument is deducted as T&. As far as I understand this is a problem when returning value.

But I could got it wrong ;-)

---

*[libs/time/timer.cpp, line 23 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXgOD6QJVcdB5TzVAMS:-KYVT2uB6_6I84gI_t25:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/0ffa8b20990830c3836f1845389b833a5d8ae63e/libs/time/timer.cpp#L23)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary><blockquote>

You can use `//` at the end of line to force specific formatting. We use that trick in few places
</blockquote></details>

Done.

---

*[libs/time/timer.cpp, line 38 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KY0QeUgtX9dXaiDGXBo:-KYVR9PKJQNpesUdy0dl:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/c5633e33364379d731545fdcef9a65b69fe2c0cf/libs/time/timer.cpp#L38)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

As those operators are declared inline and they are defined inside the source file (as opposed to header file) they will be unavailable to anyone outside this translation unit
</blockquote></details>

Done.

---

*[libs/time/include/time/timer.h, line 278 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXgTYgBcd2fGKROBpeD:-KYVTsoyKbWyes0OpB7o:b8q3wqr) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/0ffa8b20990830c3836f1845389b833a5d8ae63e/libs/time/include/time/timer.h#L278)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

did you make mark it static as well? 
</blockquote></details>

When I change it to `static constexpr TimeSpan NotificationPeriod = {TIMER_NOTIFICATION_PERIOD};` I get the following error:
```
..\build\DevBoard\lib\libtime.a(timer.cpp.obj): In function `services::time::TimeProvider::SetCurrentTime(TimePoint)':
C:\src\pwsat\src\libs\time/timer.cpp:117: undefined reference to `services::time::TimeProvider::NotificationPeriod'
collect2.exe: error: ld returned 1 exit status
src\CMakeFiles\pwsat.dir\build.make:497: recipe for target 'build/DevBoard/bin/pwsat' failed
make[2]: *** [build/DevBoard/bin/pwsat] Error 1
CMakeFiles\Makefile2:136: recipe for target 'src/CMakeFiles/pwsat.dir/all' failed
make[1]: *** [src/CMakeFiles/pwsat.dir/all] Error 2
makefile:82: recipe for target 'all' failed
make: *** [all] Error 2```
```


---

*[libs/time/include/time/timer.h, line 283 at r2](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KXgTig6TYWEzqrr7-Nx:-KY_6-_xzGIc9CGXIhZZ:b-apaz00) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/0ffa8b20990830c3836f1845389b833a5d8ae63e/libs/time/include/time/timer.h#L283)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

did you make mark it static as well? 
</blockquote></details>

See my comment above

---

*[libs/time/include/time/timer.h, line 189 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KY0Plog6ZgD3I6v8TRQ:-KYVRG528XyN6B5FWopE:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/c5633e33364379d731545fdcef9a65b69fe2c0cf/libs/time/include/time/timer.h#L189)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary><blockquote>

I think that these fields could be private
</blockquote></details>

Done.

---

*[src/mission.cpp, line 44 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KY0TeQxWJBePh94PRSJ:-KY_5ygtejMZZL2N8LCu:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/c5633e33364379d731545fdcef9a65b69fe2c0cf/src/mission.cpp#L44)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

Update the time only if it actually has been successfully obtained.
</blockquote></details>

Done.

---

*[src/obc.cpp, line 6 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KY0P-kDFAUpcGj8biQk:-KYVSvc24FOKPj3njzJv:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/c5633e33364379d731545fdcef9a65b69fe2c0cf/src/obc.cpp#L6)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary><blockquote>

Comments at the end of lines forced more readable/mergeable format - bring them back :)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53)*
<!-- Sent from Reviewable.io -->

zaq32 commented:
Reviewed 12 of 15 files at r5, 1 of 3 files at r6.
Review status: 15 of 18 files reviewed at latest revision, 9 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53)*
<!-- Sent from Reviewable.io -->

zaq32 commented:
Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[libs/time/timer.cpp, line 38 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KY0QeUgtX9dXaiDGXBo:-KYccSRKJC2c5tAhETNV:bpq3rcf) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/c5633e33364379d731545fdcef9a65b69fe2c0cf/libs/time/timer.cpp#L38)):*
<details><summary><i>Previously, mandrek44 (Mandro) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I would prefer have those methods defined in the header files as they are simple enough

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53)*
<!-- Sent from Reviewable.io -->

Novakov commented:
Reviewed 12 of 15 files at r5, 1 of 3 files at r6, 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[libs/base/Include/utils.h, line 129 at r7](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KYmbwBDvf_xt2_T-aPU:-KYmbwBDvf_xt2_T-aPV:b-h4kp0g) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/c76e17fd684403cd7ab2d3b4f33af2554363d4b4/libs/base/Include/utils.h#L129)):*
> ```objc
> template <typename T> static inline Option<std::remove_reference_t<T>> Some(T&& value)
> {
>     return Option<std::remove_reference_t<T>>::Some(std::forward<std::remove_reference_t<T>>(value));
> ```

You can add local type alias: `using U = std::remove_ference_t<T>` and simplify this statement

---

*[libs/base/Include/base/os.h, line 254 at r6](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KYbz94sGrrNlK7PkNpn:-KYbz94t1Iqe6a1qMgkq:b874rlh) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/4ad9dfdbb76d1045289f97ce789c47cc1c99706f/libs/base/Include/base/os.h#L254)):*
> ```objc
>      *
>      */
>     static OSSemaphoreHandle CreateBinarySemaphore(uint8_t semaphoreId = 0);
> ```

`std::uint8_t`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53)*
<!-- Sent from Reviewable.io -->

Novakov commented:
![](https://github.trello.services/images/mini-trello-icon.png) [[Time] Some abstraction of time](https://trello.com/c/OzyuXbqF/34-time-some-abstraction-of-time)

mandrek44 commented:
Review status: 12 of 20 files reviewed at latest revision, 3 unresolved discussions.

---

*[libs/base/Include/utils.h, line 129 at r7](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KYmbwBDvf_xt2_T-aPU:-KZSHubPoDF1nJsXKK4N:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/c76e17fd684403cd7ab2d3b4f33af2554363d4b4/libs/base/Include/utils.h#L129)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary><blockquote>

You can add local type alias: `using U = std::remove_ference_t<T>` and simplify this statement
</blockquote></details>

Done.

---

*[libs/base/Include/base/os.h, line 254 at r6](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KYbz94sGrrNlK7PkNpn:-KZSHvf_wnhvO32iiOlW:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/4ad9dfdbb76d1045289f97ce789c47cc1c99706f/libs/base/Include/base/os.h#L254)):*
<details><summary><i>Previously, Novakov (Maciej T. Nowak) wrote…</i></summary><blockquote>

`std::uint8_t`
</blockquote></details>

Done.

---

*[libs/time/timer.cpp, line 38 at r3](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-KY0QeUgtX9dXaiDGXBo:-KZSHzXbfqInewYKd-tv:b-896fix) ([raw file](https://github.com/pw-sat2/pwsat2obc/blob/c5633e33364379d731545fdcef9a65b69fe2c0cf/libs/time/timer.cpp#L38)):*
<details><summary><i>Previously, zaq32 (Łukasz Żak) wrote…</i></summary><blockquote>

I would prefer have those methods defined in the header files as they are simple enough
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53)*
<!-- Sent from Reviewable.io -->

Novakov commented:
Reviewed 8 of 9 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53)*
<!-- Sent from Reviewable.io -->

zaq32 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pw-sat2/pwsat2obc/53#-:-KZzfDMBYl55HOGtpUi0:bnfp4nl)*
<!-- Sent from Reviewable.io -->
